### PR TITLE
Notifier

### LINKED ISSUE #7549: Implement main notifier
screen shot 2016-05-27 at 17 07 10
screen shot 2016-05-27 at 17 07 54

- It should be accesible in any place of the builder.

cc @viddo

### PR DESCRIPTION
**Pull request by nobuti**
This PR closes #7549.

notifier

The code that implements the interaction above is:

```
var model = Notifier.addNotification({
  id: 'whatever',
  status: 'idle | loading | success | error',
  info: 'Uploading file.zip',
  closable: true | false,
  button: falsy | 'Label'
});

Notifier.getNotification('whatever');
Notifier.getNotification(model);
Notifier.removeNotification('whatever');
Notifier.removeNotification(model);

model.on('notification:close', callback);
model.on('notification:action', callback);
```

For now, only 3 states for a notification have been contemplated: `loading`, `success` and `error`. Taking this in mind, the error icon is not the same as the icon we have in Invision. cc @piensaenpixel .

cc @xavijam @viddo

### CODE REVIEW

**Comment by xavijam** [comment]
ping @viddo :)

**Comment by viddo** [comment]
Ditto what I wrote on the [PR of the Infobox implementation](https://github.com/CartoDB/cartodb/pull/7601#issuecomment-222958098)

**Comment by xavijam** [comment]
Yep, ^^^ let's try to encapsule a little bit all the magic. So if we implement that, we will need a way to retrieve a Notifier model in order to change its status, message or whatever:

```js
Notifier.getNotification(xxx);
```

That opens the box of how to choose the identifier, model cid, id, ...  

Or another possibility would be:

```js
…
var notificationModel = Notifier.addNotification({
  state: 'success',
  info: 'Geocoding populated_places_2',
  createActionView: function (opts) {
    …
  }
});
notificationModel.set('state', 'error');
…
```

What do you think both?

**Comment by nobuti** [comment]
I like it, in fact, I was thinking in pretty much the same thing for models we need to pass through to the infinitum like `editorModel`. Working on it, also, I'm going to get rid of the extra collection for grouping notifications.

**Comment by viddo** [comment]
@xavijam I'd like to see some high-level examples of use-cases to continue that conversation, it's not clear to me how/when these notifications should appear and change.

**Comment by xavijam** [comment]
Import a dataset example:

```js
// Import started, let's show the notification
var notificationModel = Notifier.addNotification({
  status: 'loading',
  info: 'Uploading file.zip',
  createActionView: function () {}
});

// Import has changed, it is being geocoded
notificationModel.set('info', 'Geocoding file.zip');
// or we can implement an update API
notificatonModel.updateInfo('Geocoding file.zip');

// Import has finished
notificationModel.set('status', 'succeed'); // finished, completed, succeed,... whatever
// or via public API
notificatonModel.updateStatus('succeed');

// Import has finished and there should be a button to open a window
notificationModel.set('createActionView', openSomething);
// or via public API
notificatonModel.updateAction(openSomething());

```

**Comment by xavijam** [comment]
Additional Notifier API methods I'd like to see:

```js
Notifier.addNotification({
  id: 'whatever',
  status: 'loading',
  info: 'Uploading file.zip',
  createActionView: function () {}
});

var notificationModel  = Notifier.getNotificationById('whatever'); // It will return the notification model
Notifier.removeNotification(notificationModel); // It will hide/close and remove the notification model and view
```

**Comment by viddo** [comment]
Yeah, something on those lines looks better 👍 

Also, rather than having to pass a `createActionView` I think it would make more sense to provide/document events, such as:
```js
notification.once('remove', function () { /* handle side-effect here*/ })
```

I'm skeptic about letting a developer-user set custom views in this case, from the looks of the mocks there should be a limited well-defined set of states/options for this anyway so probably better to keep that isolated and only require metadata from the outside.

**Comment by xavijam** [comment]
Ok about what you propose @viddo. I'm a bit worried too about the context menu options, not sure if we should have that because there will not be so many options, only close the notification and it is likely another one, that should be, some more info, view info or whatever. @urbanphes could you give us some light about that context menu in the "finished" notifications?

**Comment by xavijam** [comment]
@nobuti, I'd remove the context menu for the moment, we can't provide other options than close the notification. I'd add the possibility to display a 'button' with x functionality and a label.
Something like:

```js
var notificationModel = Notifier.addNotification({
  id: 'whatever',
  status: 'loading',
  info: 'Uploading file.zip',
  createActionView: function () {}
});

notificationModel.bind('buttonClicked', function () { /* whatever */ }, this);

notificationModel.set({
  status: 'completed', // or whatever
  button: _t('whatever')
});
```
not sure, what do you think?

**Comment by nobuti** [comment]
No te preocupes™

**Comment by xavijam** [comment]
:joy: -> hate you.

**Comment by nobuti** [comment]
Revamp ready. I've update the PR with the api.

**Comment by xavijam** [comment]
:clap:, I'll take a look the code first thing in the morning :)

File: lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-action-view.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+var cdb = require('cartodb.js');
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-action-view.js]: Let's change this to `CoreView` please :).

File: lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-item-view.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
+var cdb = require('cartodb.js');
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-item-view.js]: dixit

File: lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-view.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
+var cdb = require('cartodb.js');
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-view.js]: dixit

File: lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-view.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    this.clearSubViews();
+    this.$el.empty();
+    this._renderAllSubviews();
+    return this;
+  },
+
+  _initBinds: function () {
+    this.listenTo(this.collection, 'reset', this.render);
+    this.listenTo(this.collection, 'add', this._renderSubview);
+    this.listenTo(this.collection, 'remove', this._removeSubview);
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-view.js]: `this.add_related_model(this.collection);`

**Comment by xavijam** [comment]
I like it, minor comments, just the CoreView thing @viddo changed in https://github.com/CartoDB/cartodb/pull/7703. IMHO, it is ready to be 🚢 ed, @viddo any comment?

File: lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-item-view.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  },
+
+  initialize: function (opts) {
+    if (!opts.notifierModel) throw new Error('notifierModel is required');
+    if (!opts.editorModel) throw new Error('editorModel is required');
+
+    this._editorModel = opts.editorModel;
+    this.template = this.options.template || Template;
+    this._notifierModel = opts.notifierModel;
+    this.$el.attr('id', this._notifierModel.get('id'));
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-item-view.js]: model has a default id: `id: 'notifier' + _.uniqueId()`

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-item-view.js]: Argagsdgsdgsd, sorry, I didn't realize it. 👏

**Comment by alonsogarciapablo** [comment]
Looks really nice @nobuti!

File: lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-collection.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+var Backbone = require('backbone');
+var NotifierModel = require('./notifier-model');
+
+module.exports = Backbone.Collection.extend({
+  model: NotifierModel,
+  search: function (id) {
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-collection.js]: What about `findById`?

File: lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-item-view.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    this._addActionView(actionView);
+  },
+
+  _addActionView: function (view) {
+    this.$('.js-actions').append(view.render().el);
+    this.addView(view);
+  },
+
+  _closeHandler: function (action) {
+    this._notifierModel.trigger('notification:close');
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-item-view.js]: Why `notification:close` and not just `close`?

**Comment by viddo** [comment]
✨, not much to add  👍

File: lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-item-view.js (hunk 7 above)
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-item-view.js]: It's just a preference choice, I like to namespace events, I find it more readable and it helps me to locate the source. But I can change it, of course.

File: lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-collection.js (hunk 6 above)
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-collection.js]: 👍

File: lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-item-view.js (hunk 7 above)
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/editor/components/notifier/notifier-item-view.js]: We don't have a naming convention for events, so 👍

**Comment by xavijam** [comment]
Ready to go then? :)

**Comment by xavijam** [comment]
BTW, let's start using it guys ;).

**Pull request merged**